libp2p/transports/wstransport.nim

```diff
@@ -0,0 +1,199 @@
+## Nim-LibP2P
+## Copyright (c) 2021 Status Research & Development GmbH
+## Licensed under either of
+##  * Apache License, version 2.0, ([LICENSE-APACHE](LICENSE-APACHE))
+##  * MIT license ([LICENSE-MIT](LICENSE-MIT))
+## at your option.
+## This file may not be copied, modified, or distributed except according to
+## those terms.
+
+{.push raises: [Defect].}
+
+import std/[sequtils]
+import chronos, chronicles
+import transport,
+       ../errors,
+       ../wire,
+       ../multicodec,
+       ../multistream,
+       ../connmanager,
+       ../multiaddress,
+       ../stream/connection,
+       ../upgrademngrs/upgrade,
+       websock/websock
+
+logScope:
+  topics = "libp2p wstransport"
+
+export transport
+
+const
+  WsTransportTrackerName* = "libp2p.wstransport"
+
+type
+  WsStream = ref object of Connection
+    session: WSSession
+
+proc init*(T: type WsStream,
+           session: WSSession,
+           dir: Direction,
+           timeout = 10.minutes,
+           observedAddr: MultiAddress = MultiAddress()): T =
+
+  let stream = T(session: session,
+             timeout: timeout,
+             dir: dir,
+             observedAddr: observedAddr)
+  stream.initStream()
+  return stream
+
+method readOnce*(
+  s: WsStream,
+  pbytes: pointer,
+  nbytes: int):
+  Future[int] =
+  return s.session.recv(pbytes, nbytes)
+
+method write*(s: WsStream, msg: seq[byte]):
+  Future[void] {.async.} =
+  #TODO the write procs can raise LPStreamEOF
+  # BUT this is not indicated in .raises. bc of async
+  # So this function can't just return the Future of ws.send
+  # We should add {.raises.} to every write, but this will probably
+  # break many things
+  try:
+    await s.session.send(msg, Opcode.Binary)
+  except WSClosedError:
+    raise newLPStreamEOFError()
+
+method closeImpl*(s: WsStream): Future[void] {.async.} =
+  await s.session.close()
+  await procCall Connection(s).closeImpl()
+
+type
+  WsTransport* = ref object of Transport
+    httpserver: HttpServer
+    wsserver: WSServer
+    outconnections: seq[WsStream]
```

Comment:
Shouldn't be this something similar to `clients: array[Direction, seq[WSSession]]` (similar to how we handle it in the tcptransport)?

Reply:
Oups forgot the incoming connections
Though why store the incoming and outgoing in separate seq? Just for perf?

Reply:
Can't remember why it was done this way, but it really isn't that much overhead and it's always nice to have this things segregated.